### DISCUSSION TITLE
Change minimisation method in Minute

### DISCUSSION THREAD

**Discussion by thomas-kraetzschmar** [scikit-hep/pyhf#1811] [Q&A] [answered]
Dear Developers,

Is there a way to change to some of the other iminuite methods for minimization?
For example, overwrite the minimization object?

At least in the newest 2.x iminuite version, iminuite has the simplex() algorithm for minimization. That helped me a lot to get a more robust fit. But I do not know how I can set this in pyhf.

The change in the method could also help mitigate the toy-based CLs scan problems.

Also, some iminuite functions such as scan() could help users analyze their fit. It would be nice to have easy access to those functions without reprogramming pyhf functionalities.

**Comment by alexander-held** [comment]
Hi @thomas-kraetzschmar, I think it depends a bit here at what level you want to enter. The core devs can correct me, but I think if you  are operating with the `pyhf.infer` API, you cannot change the algorithm like this. You can use `return_result_obj=True` to get back the Minuit object and you can call e.g. Minos on that, but that will happen after already having run Migrad (and Hesse). I'm assuming that what you have in mind is to run Simplex before Migrad. I have not tried Scan myself, but it presumably should also work after Migrad, like Minos does (example: [part 1](https://github.com/scikit-hep/cabinetry/blob/a9b07a14fb66c30bbe12bd737ebe79c49128abc8/src/cabinetry/fit/__init__.py#L74-L95), [part 2](https://github.com/scikit-hep/cabinetry/blob/a9b07a14fb66c30bbe12bd737ebe79c49128abc8/src/cabinetry/fit/__init__.py#L260-L299)).

If you skip the `pyhf.infer` API, you can implement the minimization strategy manually, and then you have full control over the methods used. Here is [an example](https://github.com/scikit-hep/cabinetry/blob/a9b07a14fb66c30bbe12bd737ebe79c49128abc8/src/cabinetry/fit/__init__.py#L109-L198). I assume that the main gain of Simplex would be a better starting position for Migrad. If that is the case, you could also manually run a Simplex step and then again use the `pyhf.infer` API with the starting positions you obtained in your first step.

I think making the minimization approach more flexible is an interesting idea. It is not clear to me at the moment how the API for that should look. There are many potential steps with different algorithms and different options for each step that a user may in theory want to be able to use.

As a slightly tangential point, issues with Migrad minimization tend to be related to model-building, and they can usually be resolved by suitable changes to the model. Frequently that means avoiding negative yield predictions with a different modifier setup, see e.g. #1695.

**Reply by thomas-kraetzschmar** [reply]
Hi @alexander-held,

>  I assume that the main gain of Simplex would be a better starting position for Migrad. 

Yes, indeed I use your suggested strategy for what I would call top-level fits. (The initial pyhf.infer.mle.fit or even the asymptotic hypothesis test profits from this. ). 

Unfortunately, it did not work out in the case of toy-based CLs determinations. Here the toy fits fail and I have not found any good option to cure this, which forced me to reprogram this part. Hence it would be nice to set the initial minimization object or some other smart solution I  am not aware of.

**Reply by alexander-held** [reply]
I didn't correctly understand that suggestion originally, but providing a minimization object sounds like it might be a nice solution indeed!

**Reply by kratsg** [reply, marked as answer]
> Unfortunately, it did not work out in the case of toy-based CLs determinations. Here the toy fits fail and I have not found any good option to cure this, which forced me to reprogram this part. Hence it would be nice to set the initial minimization object or some other smart solution I am not aware of.

You can always write your own (custom) optimizer following `opt_minuit.py` and load it in and use it. It should be usable - or just override certain functions of the existing class implementation. If you have an idea for an API you'd like, or would like to submit a PR, go ahead. This almost looks like a feature request, but if you have a specific idea of how you'd like to see the API, let us know.
